### PR TITLE
Commit all vendored dependencies

### PR DESCRIPTION
**Pull request by metalmatze**
Reacting to the recent events with go-bindata, we agreed to commit all vendored dependencies to our repository.
I've added a new dep section to our Gopkg.toml (added in 0.4.1) to always prune dependencies. 
This simply deletes all non-go files and reduces the size from ~200M to ~30M for us.

In drone we simply run `dep status -v` now. Not sure if necessary in the end, but keeping it for now.
Something that actually checks the dependencies for outdated versions would be nice. But that's something for another PR.

### CODE REVIEW

**Comment by metalmatze** [comment]
Maybe we should simply remove `dep status -v` as it still takes 30s. That's kinda ridiculous for such a task :joy:

**Comment by mrIncompetent** [comment]
I'm not a fan of committing the vendor. It only will delay the breakage of our development to the point when we have to update our dependencies. 

And: when someone takes over a repository of a dependency and uses it for malicious purpose - this does not save us. 
Instead: This workflow will actually download the malicious code when updating. While when doing a `dep ensure -vendor-only` on every build, we know exactly when the commit hash we require, is not there anymore and the build very explicitly fails.

Also this **will** pollute our repository **and** PR's. As it is not always possible to separate vendor updates and code updates. For example: Updating the Kubernetes go-client **will** force us to update our code as well in the same PR.

I'm fine with merging this when everyone else thinks that this is ok.
@metalmatze @guusvw @alvaroaleman @realfake @p0lyn0mial

**Comment by alvaroaleman** [comment]
The advantage of this is not that it prevents us from getting maliciously altered dependencies but prevents failures due to _vanished_ dependencies.

Other advantages imho:

* A `git checkout` always results in a working copy of the repo
* Getting started is much faster

The drawback of the polluted PRs is primary a Github-Issue, not one with the concept. Other WebUIs  e.G. Bitbucket allow you to not view specific paths like the `vendor` folder. 

Also I think what @guusvw meant with splitting the vendor updates was not having dedicated PRs but having dedicated _commits_. Splitting code/dep changes into different PRs is IMHO a very bad idea.

I am in favor of doing this.

**Comment by lbb** [comment]
I think like @mrIncompetent we should not put the vendor in this repositry.
Rather we can have a separate repository with all the dependencies which we can manage independent and then set a dep alias to this repository. This allows us to better govern our dep's we use in our company, across our projects. Also this would help us to check all the licences!

**Review by alvaroaleman — DISMISSED** [review]

**Comment by metalmatze** [comment]
> It only will delay the breakage of our development to the point when we have to update our dependencies.

Yes, but it's better then letting someone else do the breakage. This way we have at least some amount of control and not the people we are vendoring from.

> And: when someone takes over a repository of a dependency and uses it for malicious purpose - this does not save us.

But now, we can actually look at the diffs produced by updating a dependency, because we can actually see what changed to the last vendored version. (I actually briefly scroll over changes when doing this)

> Instead: This workflow will actually download the malicious code when updating. While when doing a dep ensure -vendor-only on every build, we know exactly when the commit hash we require, is not there anymore and the build very explicitly fails.

I don't get your point here. We're still using dep and dep will still check the dependencies the same way as before.

> Also this will pollute our repository and PR's. As it is not always possible to separate vendor updates and code updates. For example: Updating the Kubernetes go-client will force us to update our code as well in the same PR.

Just it will pollute the repository a bit, _but_ GitHub does actually a pretty good job of not showing things in vendor expanded in PRs by default. So you'll only see that files in `vendor/` changed. Still you'd have to open the diffs yourself.
At JustWatch we had the rule, that people updating dependencies are responsible for double checking, that everything still works fine. (But we have integration tests and can automate this process, right? :joy:) I will never expect a person doing a hundred line code review to also read _all_ diffs in the changed `vendor/` files.



See the discussion with the GitHub team here: https://github.com/github/linguist/issues/3234

> The advantage of this is not that it prevents us from getting maliciously altered dependencies but prevents failures due to vanished dependencies.

Yes, this is basically everything there's to this move. Not more and not less.
Don't get leftpadded.

**Comment by alvaroaleman** [comment]
> Rather we can have a separate repository with all the dependencies which we can manage independent and then set a dep alias to this repository. This allows us to better govern our dep's we use in our company, across our projects.

This will fail the instant projects use different versions of the same dependency.

**Comment by metalmatze** [comment]
What Go really needs in a perfect world is a mirror of all repositories and their releases. 
See @sdboyer's comment on this project: https://github.com/whyrusleeping/gx/issues/141

Until Go comes up with a better way, we should simply commit dependencies.

**Comment by mrIncompetent** [comment]
> >  And: when someone takes over a repository of a dependency and uses it for malicious purpose - this does not save us.
> 
> But now, we can actually look at the diffs produced by updating a dependency, because we can actually see what changed to the last vendored version. (I actually briefly scroll over changes when doing this)
>
> > Instead: This workflow will actually download the malicious code when updating. While when doing a dep ensure -vendor-only on every build, we know exactly when the commit hash we require, is not there anymore and the build very explicitly fails.
> 
> I don't get your point here. We're still using dep and dep will still check the dependencies the same way as before.

Scenario: Someone deletes his repository and an attacker takes it over, commits an old version of the previous hosted code & then commits some malicious code.

**When not committing our code**:
Our CI would fail, as `dep ensure -vendor-only` dep will fails - the commit hash from the Gopkg.lock does not exist anymore.

**When committing our code**:
Our CI works without any issues.
At some point i update the dependencies using `dep update` -> dep will download the newest version of the dependency -> the version with the malicious code.

At least i doubt that we'll detect that.

**Comment by metalmatze** [comment]
> **When not committing our code:**
> Our CI would fail, as dep ensure -vendor-only dep will fails - the commit hash from the Gopkg.lock does not exist anymore.

I think the person who took over the user and go-bindata repository simply uploaded the same repository. So hashes didn't change.

Someone would eventually update our deps and fetch a new go-bindata version, which could then contain malicious code. But we will never know, because the repository seems to be the correct one and the new commits, made to that repository, are just accepted by dep. We have no way of verifying these unless going to the GitHub repository and looking through them there (or clone to local machine and do the same).

> **When committing our code:**
> Our CI works without any issues.
> At some point i update the dependencies using dep update -> dep will download the newest version of the dependency -> the version with the malicious code.

Yes. Our CI works without any issues, even when somebody deletes their user or repository.
Once we notice that the repository, we depend on, is gone we have all the time in the world to switch to a different one or can work on an upstream fix. That wouldn't be possible when not committing `vendor/`.

> dep will download the newest version of the dependency -> the version with the malicious code.

Yes, this will still happen. Nevertheless we now have the ability to see exactly what changed in a git diff. It should be the responsibility of the developer, updating a dependency, to verify the new code and only then commit the new changes in the vendored dependencies.
At least we now have the opportunity to check ourselves.

To summarize:

:+1: We don't break when somebody decides to remove our dependencies.
:+1: We have the control to update our dependencies when we want to and double check the changes.
:-1: It clutters the git history

**Comment by mrIncompetent** [comment]
@metalmatze True, haven't though of that.

We should probably add a hack-script for updating dependencies. (So we don't forget the `dep prune`)

**Comment by metalmatze** [comment]
As of dep version 0.4.1 pruning is done automatically by with `dep ensure`. You can't forget it and we don't need a script.

I've added the section for dep on how to prune `vendor/`.
https://github.com/kubermatic/kubermatic/pull/732/files#diff-999f7befaa72fef803f7bc585cc68f72R3

**Comment by mrIncompetent** [comment]
I'm fine with merging

**Comment by guusvw** [comment]
:+1:

**Review by guusvw — DISMISSED** [review]

**Comment by p0lyn0mial** [comment]
It's hard not to disagree with @mrIncompetent  that PR's diffs can be difficult to deal with. On the other hand including `vendor` in the repository decouples us from the upstream which seems to be tempting. Tradeoffs :) I think we can give it a try as we can always restore to the prevous state.

@metalmatze  it is still required to run `dep ensure` ? (https://github.com/kubermatic/kubermatic/blob/master/api/Makefile#L56)

And I know that the documentations says "it is never a bad thing to run `dep ensure`" :)
I thinkg that having `vendor` in the reposiotry doesn't require to run any additional commands.

**Review by mrIncompetent — DISMISSED** [review]

**Comment by metalmatze** [comment]
Yes, @p0lyn0mial. It's a tradeoff for sure!

We don't really need to run `vendor` before running `test` or `install`, that is true. I will update the PR. I'll leave the `make vendor` target so you can still run this, if needed.

**Review by p0lyn0mial — DISMISSED** [review]

**Comment by alvaroaleman** [comment]
@metalmatze Can you fix this and get it merged? :)
